File: app/src/main/java/com/example/myapplication/MainActivity.java
package com.example.myapplication;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import androidx.activity.ComponentActivity;
import androidx.lifecycle.ViewModelProvider;

import com.example.myapplication.databinding.ActivityMainBinding;

public class MainActivity extends ComponentActivity {

    MyViewModel myViewModel;

//    TextView txtShowCounter;
//    Button btnAddCounter;

    ActivityMainBinding activityMainBinding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        setContentView(R.layout.activity_main);
        activityMainBinding = ActivityMainBinding.inflate(getLayoutInflater());

        myViewModel = new ViewModelProvider(this).get(MyViewModel.class);
//        txtShowCounter = findViewById(R.id.txt_show_counter);
//        btnAddCounter = findViewById(R.id.btn_add_counter);

        showCounter();
//        btnAddCounter.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                myViewModel.counter++;
//                showCounter();
//            }
//        });
//        activityMainBinding.btnAddCounter.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                myViewModel.counter++;
//                showCounter();
//            }
//        });
        //include layout的情况。
        activityMainBinding.includes.btnAddCounter.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                myViewModel.counter++;
                showCounter();
            }
        });

        setContentView(activityMainBinding.getRoot());
    }

    private void showCounter() {
//        txtShowCounter.setText(
//                String.format(
//                        getResources().getString(R.string.counter_value),
//                        myViewModel.counter)
//        );
        activityMainBinding.txtShowCounter.setText(
                String.format(
                        getResources().getString(R.string.counter_value),
                        myViewModel.counter)
        );
    }
}

File: databindingdemo/src/main/java/com/example/databindingdemo/MainActivity.kt
package com.example.databindingdemo

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.activity.viewModels
import androidx.databinding.DataBindingUtil
import com.example.databindingdemo.databinding.ActivityMainBindingImpl
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val userVM: UserViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        setContentView(R.layout.activity_main)
        val dataBinding: ActivityMainBindingImpl = DataBindingUtil.setContentView(this, R.layout.activity_main)
        dataBinding.userViewModel = userVM

        dataBinding.button.setOnClickListener {
            if (userVM.firstName.get().isNullOrEmpty() || userVM.lastName.get().isNullOrEmpty()
                    || userVM.firstName.get().equals("long") || userVM.lastName.get().equals("click")) {
                userVM.firstName.set("qin")
                userVM.lastName.set("peng")
            }
        }

        // kotlin support.
        button.setOnLongClickListener {
            userVM.firstName.set("long")
            userVM.lastName.set("click")
            true
        }
    }
}
